compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts

```diff
@@ -516,6 +519,68 @@ DEFAULT_GLOBALS.set(
   addObject(DEFAULT_SHAPES, 'global', TYPED_GLOBALS),
 );
 
+export function installCustomGlobals(
+  globals: GlobalRegistry,
+  registry: ShapeRegistry,
+  custom: Array<[string, GlobalType]>,
+): void {
+  const refMap: Map<number, BuiltInType> = new Map();
+  const toSet: Map<number, (ty: BuiltInType) => void> = new Map();
+  function installShape(name: string, type: GlobalType): BuiltInType {
+    const props: Array<[string, BuiltInType]> = type.properties.map(
+      ([name, type]) => [name, installShape(name, type)],
+    );
+
+    let ty;
+    if (type.fn == null) {
+      ty = addObject(registry, name, props);
+    } else {
+      const func: Omit<FunctionSignature, 'hookKind'> = {
+        positionalParams: type.fn.positionalParams,
+        restParam: type.fn.restParam,
+        calleeEffect: type.fn.calleeEffect,
+        returnValueKind: type.fn.returnValueKind,
+        returnType: {kind: 'Primitive'}, // Placeholder, mutated below
+      };
+
+      if (type.fn.returnType == null) {
+        func.returnType = {kind: 'Poly'};
+      } else if (type.fn.returnType === 'array') {
+        func.returnType = {kind: 'Object', shapeId: BuiltInArrayId};
+      } else if (refMap.has(type.fn.returnType)) {
+        func.returnType = refMap.get(type.fn.returnType)!;
+      } else {
+        const cur = toSet.get(type.fn.returnType);
+        toSet.set(type.fn.returnType, (ty: BuiltInType) => {
+          cur?.(ty);
+          func.returnType = ty;
+        });
+      }
```

Comment:
does this allow custom types to refer to builtin types? would be nice if that was the case (helper function that returns a ref, for example)

Reply:
Yeah! Just need to add 'ref', etc to `TypeReferenceSchema` and then link it with the builtin ref shape.